### PR TITLE
store/tikv: mantain lastTS in pd oracle.

### PR DESCRIPTION
**Pull request by disksing**
@coocood

### CODE REVIEW

File: store/tikv/oracle/oracles/pd.go [hunk 1]
```diff
@@ -28,30 +29,55 @@ const slowDist = 30 * time.Millisecond
...
+		o.Close()
+		return nil, errors.Trace(err)
 	}
+	return o, nil
 }
 
 // IsExpired returns whether lockTs+TTL is expired, both are ms.
-func (t *pdOracle) IsExpired(lockTs, TTL uint64) (bool, error) {
-	physical, _, err := t.c.GetTS()
+func (o *pdOracle) IsExpired(lockTs, TTL uint64) (bool, error) {
```
Comment (coocood) [inline store/tikv/oracle/oracles/pd.go]: The returned error can be removed.

File: store/tikv/oracle/oracles/pd.go [hunk 2]
```diff
@@ -28,30 +29,55 @@ const slowDist = 30 * time.Millisecond
...
+	quit    chan struct{}
+	updated chan struct{}
 }
 
 // NewPdOracle create an Oracle that uses a pd client source.
 // Refer https://github.com/pingcap/pd/blob/master/pd-client/client.go for more details.
-func NewPdOracle(pdClient pd.Client) oracle.Oracle {
-	return &pdOracle{
-		c: pdClient,
+func NewPdOracle(pdClient pd.Client, updateInterval time.Duration) (oracle.Oracle, error) {
```
Comment (coocood) [inline store/tikv/oracle/oracles/pd.go]: Add comment about `updateInterval`.

File: store/tikv/oracle/oracles/pd.go (hunk 1 above)
Comment (coocood) [inline store/tikv/oracle/oracles/pd.go]: Add comment explain that this function may return false negative result as lockTs is compared to `lastTs`.

Comment (disksing) [inline store/tikv/oracle/oracles/pd.go]: I think it's okay to keep the error. The signature is defined by the Oracle interface, and 'not return an error' is implementation specific.

**Comment by shenli** [comment]
LGTM

**Comment by disksing** [comment]
PTAL @coocood @shenli

**Comment by coocood** [comment]
LGTM

File: store/tikv/kv.go [hunk 3]
```diff
@@ -64,29 +64,40 @@ func (d Driver) Open(path string) (kv.Storage, error) {
...
+	s, err := newTikvStore(uuid, &codecPDClient{pdCli}, newRPCClient())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
 	mc.cache[uuid] = s
 	return s, nil
 }
 
+// update oracle's lastTS every 2000ms.
+var oracleUpdateInterval = 2000
```
Comment (zimulala) [inline store/tikv/kv.go]: 2000 -> 2000 * time.Millisecond ?

**Pull request merged**